CSharp/Library/Microsoft.Bot.Builder/Dialogs/PromptRecognizers.cs

```diff
@@ -0,0 +1,460 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using Microsoft.Bot.Connector;
+using System.Resources;
+using System.Globalization;
+
+namespace Microsoft.Bot.Builder.Dialogs
+{
+    public class RecognizeEntity<T>
+    {
+        public T Entity { get; set; }
+        public double Score { get; set; }
+    }
+    public interface IPromptRecognizeNumbersOptions
+    {
+        int? MinValue { get; set; }
+        int? MaxValue { get; set; }
+        bool? IntergerOnly { get; set; }
+    }
+
+    public class PromptRecognizeNumbersOptions : IPromptRecognizeNumbersOptions
+    {
+        /// <summary>
+        /// (Optional) Minimum value allowed.
+        /// </summary>
+        public int? MinValue { get; set; }
+        /// <summary>
+        /// (Optional) Maximum value allowed.
+        /// </summary>
+        public int? MaxValue { get; set; }
+        /// <summary>
+        /// (Optional) If true, then only integers will be recognized.
+        /// </summary>
+        public bool? IntergerOnly { get; set; }
+    }
+
+    public interface IPromptRecognizeValuesOptions
+    {
+        bool? AllowPartialMatches { get; set; }
+        int? MaxTokenDistance { get; set; }
+    }
+
+    public interface IPromptRecognizeChoicesOptions : IPromptRecognizeValuesOptions
+    {
+        bool? ExcludeValue { get; set; }
+        bool? ExcludeAction { get; set; }
+    }
+
+    public class PromptRecognizeChoicesOptions : IPromptRecognizeChoicesOptions
+    {
+        /// <summary>
+        /// (Optional) If true, the choices value will NOT be recognized over.
+        /// </summary>
+        public bool? ExcludeValue { get; set; }
+        /// <summary>
+        /// (Optional) If true, the choices action will NOT be recognized over.
+        /// </summary>
+        public bool? ExcludeAction { get; set; }
+        /// <summary>
+        /// (Optional) if true, then only some of the tokens in a value need to exist to be considered a match.The default value is "false".
+        /// </summary>
+        public bool? AllowPartialMatches { get; set; }
+        /// <summary>
+        /// (Optional) maximum tokens allowed between two matched tokens in the utterance.So with
+        /// a max distance of 2 the value "second last" would match the utternace "second from the last"
+        /// but it wouldn't match "Wait a second. That's not the last one is it?". 
+        /// The default value is "2".  
+        /// </summary>
+        public int? MaxTokenDistance { get; set; }
+    }
+    
+    public class LocalizedCache<T>
+    {
+        public string Id { get; set; }
+        public IDictionary<string, T> Locales { get; set; }
+
+        public LocalizedCache()
+        {
+            this.Locales = new Dictionary<string, T>();
```

Comment:
>Dictionary [](start = 31, length = 10)

should this be a ConcurrentDictionary? different threads with same ambient thread culture might try to add to this dictionary.